planner/core/stats.go

```diff
@@ -150,12 +151,109 @@ func (ds *DataSource) DeriveStats(childStats []*property.StatsInfo) (*property.S
 			break
 		}
 	}
+	// consider the IndexMergePath. Now, we just generate `IndexMergePath` in DNF case.
+	if len(ds.pushedDownConds) > 0 && len(ds.possibleAccessPaths) > 2 && ds.ctx.GetSessionVars().EnableIndexMerge {
+		needConsiderIndexMerge := true
+		for i := 1; i < len(ds.possibleAccessPaths); i++ {
+			if len(ds.possibleAccessPaths[i].accessConds) != 0 {
+				needConsiderIndexMerge = false
+				break
+			}
+		}
+		if needConsiderIndexMerge {
+			ds.generateIndexMergeOrPaths()
+		}
+	}
 	if ds.ctx.GetSessionVars().OptimizerSelectivityLevel >= 1 {
 		ds.stats.HistColl = finalHist
 	}
 	return ds.stats, nil
 }
 
+// getIndexMergeOrPath generates all possible IndexMergeOrPaths.
+func (ds *DataSource) generateIndexMergeOrPaths() {
+	usedIndexCount := len(ds.possibleAccessPaths) - 1
+	for i, cond := range ds.pushedDownConds {
+		var ixMergePartialIxPaths = make([]*accessPath, 0, usedIndexCount)
+
+		sf, ok := cond.(*expression.ScalarFunction)
+		if !ok || sf.FuncName.L != ast.LogicOr {
+			continue
+		}
+		dnfItems := expression.FlattenDNFConditions(sf)
+		for _, item := range dnfItems {
+			var itemPaths []*accessPath
+			cnfItems := expression.SplitCNFItems(item)
+			itemPaths = ds.accessPathsForConds(cnfItems, usedIndexCount)
+			if len(itemPaths) == 0 {
+				ixMergePartialIxPaths = nil
+				break
+			}
+			partialPath := ds.buildIndexMergePartialPath(itemPaths)
+			if partialPath == nil {
+				ixMergePartialIxPaths = nil
+				break
+			}
+			ixMergePartialIxPaths = append(ixMergePartialIxPaths, partialPath)
+		}
+		if len(ixMergePartialIxPaths) > 1 {
+			possiblePath := ds.buildIndexMergeOrPath(ixMergePartialIxPaths, i)
+			if possiblePath != nil {
+				ds.possibleAccessPaths = append(ds.possibleAccessPaths, possiblePath)
+			}
+		}
+	}
+}
+
+// accessPathsForConds generates all possible index paths for conditions.
+func (ds *DataSource) accessPathsForConds(conditions []expression.Expression, usedIndexCount int) []*accessPath {
+	var results = make([]*accessPath, 0, usedIndexCount)
+	for i := 1; i <= usedIndexCount; i++ {
+		path := &accessPath{index: ds.possibleAccessPaths[i].index}
+		_, err := ds.deriveIndexPathStats(path, conditions)
+		if err != nil {
+			return nil
+		}
+		// if accessConds is empty or tableFilter is not empty, we ignore the access path.
+		// now these conditions are too strict.
+		// for example, a sql `select * from t where a > 1 or (b < 2 and c >3)` and table `t` with indexes on a and b separately.
+		// we can generate a `IndexMergePath` with table filter `a > 1 or (b < 2 and c >3)`.
+		// TODO: solve the above case
+		if len(path.tableFilters) > 0 || len(path.accessConds) == 0 {
```

Comment:
These two conditions can be replaced by `len(path.accessConds) == len(conditions)`?